ports/zephyr/machine_pin.c

```diff
@@ -151,6 +191,67 @@ STATIC mp_obj_t machine_pin_on(mp_obj_t self_in) {
 }
 STATIC MP_DEFINE_CONST_FUN_OBJ_1(machine_pin_on_obj, machine_pin_on);
 
+// pin.irq(handler=None, trigger=IRQ_FALLING|IRQ_RISING, hard=False)
+STATIC mp_obj_t machine_pin_irq(size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_handler, ARG_trigger, ARG_hard };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_handler, MP_ARG_OBJ, {.u_rom_obj = MP_ROM_NONE} },
+        { MP_QSTR_trigger, MP_ARG_INT, {.u_int = GPIO_INT_EDGE_BOTH} },
+        { MP_QSTR_hard, MP_ARG_BOOL, {.u_bool = false} },
+    };
+    machine_pin_obj_t *self = MP_OBJ_TO_PTR(pos_args[0]);
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    if (self->irq == NULL) {
+        machine_pin_irq_obj_t *irq;
+        for (irq = MP_STATE_PORT(machine_pin_irq_list); irq != NULL; irq = irq->next) {
+            machine_pin_obj_t *irq_pin = MP_OBJ_TO_PTR(irq->base.parent);
+            if (irq_pin->port == self->port && irq_pin->pin == self->pin) {
+                break;
+            }
+        }
+        if (irq == NULL) {
+            irq = m_new_obj(machine_pin_irq_obj_t);
+            irq->base.base.type = &mp_irq_type;
+            irq->base.methods = (mp_irq_methods_t *)&machine_pin_irq_methods;
+            irq->base.parent = MP_OBJ_FROM_PTR(self);
+            irq->base.handler = mp_const_none;
+            irq->base.ishard = false;
+            irq->next = MP_STATE_PORT(machine_pin_irq_list);
+            gpio_init_callback(&irq->callback, gpio_callback_handler, 0);
+            int ret = gpio_add_callback(self->port, &irq->callback);
+            if (ret != 0) {
+                mp_raise_OSError(-ret);
+            }
+            MP_STATE_PORT(machine_pin_irq_list) = irq;
+        }
+        self->irq = irq;
+    }
+
+    if (n_args > 1 || kw_args->used != 0) {
+        // configure irq
+        self->irq->callback.pin_mask &= ~BIT(self->pin);
```

Comment:
We shouldn't be modifying the callback structure internals; `pin_mask` should get set by `gpio_init_callback`.

Reply:
Ok.  I was a bit uneasy about modifying it directly but the comment in gpio.h did say "Such pin_mask can be modified whenever necessary by the owner".

Anyway, I changed it now and made it a bit simpler so `gpio_init_callback()` is only called once with `BIT(self->pin)`, and `gpio_pin_interrupt_configure()` is used exclusively to disable/enable the IRQ.